### ISSUE TITLE
Debounce and caching

### ISSUE THREAD

**Issue by tombh** [openlawlibrary/pygls#283]
Just wondering out loud whether debouncing and caching is something that Pygls could support?

The classic example is completion requests, they're potentially sent on every keystroke, which, depending on the implementation could be quite expensive, even before considering a separate `completionItem/resolve` step. One solution is to debounce the requests, say for a few 100ms, but that still requires sending _something_ back to the client for each request, hence the need for also providing a cache.

Debouncing and caching aren't the hardest things in the world, but still they can be a bit fiddly to get all the little details right. What's more, each LSP feature has its own debounce and caching requirements. Say for instance, diagnostic requests don't need an immediate response, so don't need caching, but they should defer the most recent request to the future in order to be published after the debounce limit. Formatters probably don't need either debounce or caching. And so on.

So could Pygls itself be a good place to at least offer APIs and recommended approaches?

**Comment by llllvvuu** [comment]
A nice API for this could be via the decorator:

```python
@ls.feature(TEXT_DOCUMENT_DID_CHANGE, debounce_interval=100)
def did_change(params: DidChangeTextDocumentParams):
   ...
```

IDK about caching, maybe it can be up to the client side to decide to show stale content

**Comment by llllvvuu** [comment]
> but that still requires sending _something_ back to the client for each request

There's also the semantic sense in which this is true of the current API: expecting request handlers to be functions that return a value rather than do a thing and return nothing, means that you have to have a one-to-one relationship between requests and responses (i.e. no dropped requests or streamed responses). This doesn't seem easy to route around.

**Comment by tombh** [comment]
I like the dedicated decorator idea.

You had more comments before about caching I think? My assumption is that without caching the editor might flash diagnostics on and off. If the editor receives an empty diagnostics lists then it assumes that everything was fixed right?

Ah yes, I didn't think about how editors might have their own debouncing, good point.

> So possibly server-side debounce interval should be configurable if the user doesn't feel double debounce is necessary.

Do you mean some other config apart from the decorator args in `@debounce_and_throttle(100, 100)`?

**Comment by llllvvuu** [comment]
> If the editor receives an empty diagnostics lists then it assumes that everything was fixed right?

For diagnostics specifically, those are one-way notifications rather than request/response. So the server can just not send a notification. In fact pygls handler for `textDocument/didChange` expects None I believe.

> Do you mean some other config apart from the decorator args in @debounce_and_throttle(100, 100)?

Nah you're right, just passing a variable in instead of hardcoded 100 works. That's [what I ended up doing](https://github.com/Instagram/Fixit/pull/392).

**Comment by llllvvuu** [comment]
> Ah yes, I didn't think about how editors might have their own debouncing, good point.

Yeah we might be in uncharted territory here, I suspect language servers tend to debounce diagnostic notifications but not request/response, assuming that the client side will handle this / the server will never get ddos'ed / all the request/response-based methods are fast enough to not need it, unlike linters (completions might actually be the only borderline case)

**Comment by llllvvuu** [comment]
Another reason to have it be a separate utility is that in many cases you probably want to debounce some subroutine and not the handler itself:

```python
@ls.feature(TEXT_DOCUMENT_DID_CHANGE)
def did_change(params: DidChangeTextDocumentParams):
   update_index(params.textDocument)  # must be fresh
   debounce(0.2)(validate(params.textDocument))  # slower
```

or if validate should always be debounced:

```python
@ls.feature(TEXT_DOCUMENT_DID_CHANGE)
def did_change(params: DidChangeTextDocumentParams):
   update_index(params.textDocument)  # must be fresh
   validate(params.textDocument)  # slower

@debounce(0.2)
def validate(textDocument):
  ...
```

**Comment by tombh** [comment]
> For diagnostics specifically, those are one-way notifications rather than request/response. So the server can just not send a notification.

Ah yes, of course.

> instead of responding to request N with request N-1's return value, respond to request N with request N+1's return value

Such an elegant tweak, nice.

Your Fixit project looks interesting. I've been working on a framework too for making it easier to write custom LSP functionality, here's my debounce code: https://github.com/tombh/super-glass-lsp/blob/main/super_glass_lsp/lsp/custom/features/_debounce.py It's been a while since I worked on it, and I notice that I do debounce formatting requests, even though in my first comment above I suggest that formatters don't need debounce, and you're saying that completion requests are probably the only thing that needs debouncing. Thinking about it now, I do feel like debounce is useful for formatting requests. Can you think of a reason why it isn't?

> assuming that the client side will handle this / the server will never get ddos'ed 

One of the reasons I'd like to include debouncing in Pygls is that it provides a mechanism for server authors to be both prompted to think about debouncing where it's relevant (eg; we could provide a default `@debounce(0)` on relevant features) and to encourage a way to provide a recommended debounce time that's tailored to the specific tool/algorithm generating the response. The latter being something that a generic client-side debounce wouldn't know about.

```python
@debounce(0.2)
def validate(textDocument):
```
Yeah that's exactly how I'd imagine it.
